### ISSUE TITLE
Questions on permissions requested by the RiMusic app

### ISSUE THREAD

**Issue by IzzySoft** [fast4x/RiMusic#400]
### Steps to reproduce the bug

Look at the permissions (either in Android's app list, or [at the `AndroidManifest.xml`](https://github.com/fast4x/RiMusic/blob/master/app/src/main/AndroidManifest.xml#L16)

### Expected behavior

According to the app description, RiMusic is supposed to **play** audio – but not to **record** it, so requesting this permission is quite "unexpected behavior" :wink: and it should rather **not** be requested.

### Actual behavior

The permission is declared in the Manifest.

### Screenshots/Screen recordings

n/a

### Logs

n/a

### RiMusic version

0.6.17

### Android version

any

### Additional information

Not complaining, but this looks weird and might raise suspicions. So could you please remove this permission – or if it's needed indeed, clarify that in the app description (Fastlane, Readme etc)?

**Comment by IzzySoft** [comment]
PS: I see this issue got the number 400, so HTTP respond code seems to concur this is a "400 Bad Request" :see_no_evil:

**Comment by fast4x** [comment]
Hi Izzy, permission is only required to use the visualizer. By default the visualizer is disabled in the settings. When the user activates it because he wants to use it, permission is required and the user is free not to grant it.

I immediately add the details in the readme, fastlane and wiki FAQ

**Comment by ikanakova** [comment]
@fast4x I think it would be great to list all the permissions the app requires and why.

**Comment by fast4x** [comment]
> @fast4x I think it would be great to list all the permissions the app requires and why.

Good idea

**Comment by ikanakova** [comment]
I think that's enough to be listed in the fastlane and FAQs. And mention in the readme with a link to the wiki.

**Comment by fast4x** [comment]
Yes

**Comment by fast4x** [comment]
All updated, readme and fastlane.
Here faqs page with all permissione
https://github.com/fast4x/RiMusic/wiki/FAQs#-what-permissions-are-used

**Comment by IzzySoft** [comment]
First thanks for the quick action! I still wonder:

> permission is only required to use the visualizer

Why does a **visualizer** require access to the microphone? It's not called "audioliser". Visuals are to watch, not to "talk to" :flushed: Can you help me understand what I miss there? I guess I'm not the only one "with a knot in the head" there, so maybe a hint in the Wiki (yay, great idea to list all permissions there, @ikanakova!) would be good. Oh, and maybe a hint in the Fastlane `full_description` that permissions are explained in the (link) Wiki?

**Comment by fast4x** [comment]
I added the link to the wiki in the fastlane, I thought I already did.

According to the official visualizer documentation:
"The Visualizer class enables application to retrieve part of the currently playing audio for visualization purpose. It is not an audio recording interface and only returns partial and low quality audio content. However, to protect privacy of certain audio data (e.g voice mail) the use of the visualizer requires the permission android.permission.RECORD_AUDIO."

https://developer.android.com/reference/android/media/audiofx/Visualizer

**Comment by IzzySoft** [comment]
Hm, K. In a way, it is "recording", yeah. Confusing though, as it belongs to the permission group "Microphone" ("Direct access to the microphone to record audio."), [see here](http://androidpermissions.com/permission/android.permission.RECORD_AUDIO). But it seems it was removed from that group at some point, and [now is "undefined"](https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/res/AndroidManifest.xml#L1575). Well, "the ways of Google are inscrutable"…

Thanks a lot! I've added it to the allow-list for your app now, pointing to the developer link you just quoted. Thanks also for updating the full description in fastlane once more! All done then here, so I "close the door" now. Should there be any questions left for me on your end, just give me a ping :smile:

**Comment by fast4x** [comment]
Ping for you 😂
Thanks for your attention, I like this kind of interaction. I have a question about the F-Droid build process for you if you can give me an indication.

It’s happening that I publish the new version on GitHub, load the apk built with the same committed source, but F-Droid does not build from this tag because it finds differences. This thing happens from a couple of versions. The tag and apk are correct, I don’t understand.

F-Droid creates from the tag, I do not understand why it finds differences.

**Comment by ikanakova** [comment]
> https://developer.android.com/reference/android/media/audiofx/Visualizer

I've added this link to the wiki as well to make it clear that this is an Android invention.

**Comment by IzzySoft** [comment]
@fast4x which differences? See [HOWTO: diff & fix APKs for Reproducible Builds](https://gitlab.com/fdroid/wiki/-/wikis/HOWTO:-diff-&-fix-APKs-for-Reproducible-Builds). There can be any of several reasons. Example: You build on Windows, which generates files with `\r\n`  as EOL. F-Droid builds on Linux, which uses `\n`. Same source, different results. Other examples behind the link – including how to solve them (if possible).

**Comment by fast4x** [comment]
Perfect! I’ll look at it right away, thank you!

**Comment by fast4x** [comment]
#397 it's the last reported difference I’m trying to figure out.

**Comment by IzzySoft** [comment]
Help is on its way, see my comment there (I was performing those RB tests often while I was active at F-Droid, hence I have the proper tools at hand – that `diff.html`  is just "too much input" to be helpful for someone not accustomed with the task).

**Comment by IzzySoft** [comment]
OK, next batch (with all checks active now in my scanner):

    ! repo/it.fast4x.rimusic_14.apk declares risky permission(s): android.permission.READ_EXTERNAL_STORAGE android.permission.READ_MEDIA_AUDIO

I guess we can add `READ_MEDIA_AUDIO` to the allow-list straight away, as your music app has a good reason to want that (play audio stored there). But why `READ_EXTERNAL_STORAGE`?  Minimum Android version needed to use your app is Android 5, which already has SAF (Storage Access Framework) available, so there should be no need for the older access methods. Or do I miss something here?

Thanks ahead once more for shedding light!

**Comment by fast4x** [comment]
Hi Izzy, you’re right, actually that permission (READ_EXTERNAL...) was already there since i forked ViMusic.

The app also requires WRITE_EXTERNAL_STORAGE permission which is used up to Android 10 to set custom cache.

READ_MEDIA_AUDIO it's used for read local songs...

I always thank you for your kindness.

**Comment by fast4x** [comment]
Confirmed, i have removed READ_EXTERNAL_STORAGE and WRITE_EXTERNAL_STORAGE not needed.

**Comment by fast4x** [comment]
Also updated wiki page https://github.com/fast4x/RiMusic/wiki/FAQs#-what-permissions-are-used

**Comment by IzzySoft** [comment]
Thanks a lot! Added `READ_MEDIA_AUDIO` to your app's allow-list. The other two should then be gone with the next release and thus need now allow-listing.

**Comment by IzzySoft** [comment]
@fast4x looks like that permission list needs an update. Today my scanner reported additional things:

* android.permission.REQUEST_INSTALL_PACKAGES: what for does a music player need to install apps?
* usesCleartextTraffic: where are unencrypted connections established to?

**Comment by ikanakova** [comment]
> @fast4x looks like that permission list needs an update. Today my scanner reported additional things:
> 
> android.permission.REQUEST_INSTALL_PACKAGES: what for does a music player need to install apps?
> usesCleartextTraffic: where are unencrypted connections established to?

I think this is related to issue #555

**Comment by fast4x** [comment]
Hi @IzzySoft, 
from this version check update feature include download and install package from GitHub, when user click Update button. #555 

Library in use is AppUpdate, look at this GitHub link https://github.com/azhon/AppUpdate

All connections to GitHub or YT/YTM from app are in https, not have connection unencrypted.

**Comment by fast4x** [comment]
Updated permissions list in the wiki page
https://github.com/fast4x/RiMusic/wiki/FAQs#-what-permissions-are-used

**Comment by IzzySoft** [comment]
> check update feature include download and install package from GitHub

Thanks @fast4x! So the app now has a self-updater? You're aware that violates the inclusion policy of F-Droid and my repo as well (as in both cases, the additional screening applied by the repos is skipped)?

> when user click Update button

So it's kinda opt-in then (update check only happens when that button is pressed)? Is it clear to someone not having followed discussions here that updates then happen directly from Github – and are the implications (circumventing the extra security provided by F-Droid/my repo) clearly pointed out?

> All connections to GitHub or YT/YTM from app are in https, not have connection unencrypted.

Then why does the app have the `usesCleartextTraffic` set?

```xml
<application android:theme="@7F0F011C" android:label="RiMusic" android:icon="@7F0D0001"
 android:name="it.vfsfitvnm.vimusic.MainApplication" android:configChanges="0x40007FFF"
 android:allowBackup="true" android:hardwareAccelerated="true" android:supportsRtl="true"
 android:banner="@7F0D0000" android:extractNativeLibs="true" android:usesCleartextTraffic="true"
```

Maybe some dependency merged that in, and you could "override" it?

**Comment by DarkCrypt** [comment]
@fast4x 
My apologies friend, I didn't mean for you to do pointless work.

@IzzySoft 
I'm the one who requested the in-app updater feature. I didn't realize this was going to cause such conflict. This is extremely upsetting.

**Comment by fast4x** [comment]
> > check update feature include download and install package from GitHub
> 
> Thanks @fast4x! So the app now has a self-updater? You're aware that violates the inclusion policy of F-Droid and my repo as well (as in both cases, the additional screening applied by the repos is skipped)?
> 
> > when user click Update button
> 
> So it's kinda opt-in then (update check only happens when that button is pressed)? Is it clear to someone not having followed discussions here that updates then happen directly from Github – and are the implications (circumventing the extra security provided by F-Droid/my repo) clearly pointed out?
> 
> > All connections to GitHub or YT/YTM from app are in https, not have connection unencrypted.
> 
> Then why does the app have the `usesCleartextTraffic` set?
> 
> ```xml
> <application android:theme="@7F0F011C" android:label="RiMusic" android:icon="@7F0D0001"
>  android:name="it.vfsfitvnm.vimusic.MainApplication" android:configChanges="0x40007FFF"
>  android:allowBackup="true" android:hardwareAccelerated="true" android:supportsRtl="true"
>  android:banner="@7F0D0000" android:extractNativeLibs="true" android:usesCleartextTraffic="true"
> ```
> 
> Maybe some dependency merged that in, and you could "override" it?

@IzzySoft 
Related to cleartext i had it enabled for debugging, I forgot to disable it. I'm afraid, disable it and rebuild.

Check update, check if there is an update on startup, if there is shows the user a message with an update button. Only if the user clicks Update start download and installation which still requires the user to confirm further.

This violates FDroid policy? Aigoo...

**Comment by fast4x** [comment]
@IzzySoft 
https://gitlab.com/fdroid/fdroiddata/-/issues/3110#note_1613431934

Could I make it configurable from the settings if I use the simple check update or the new one that also allows you to download and launch the installation?

So by default I would have the verification as before, the user could activate the new mode from the settings, in both cases the download is still from github as before.

**Comment by fast4x** [comment]
@IzzySoft 
`The software must not download additional executable binary files (e.g. addons, auto-updates, etc.) without explicit user consent. Consent means it needs to be opt-in (it must not be harder to decline than to accept or presented in a way users are likely to press accept without reading) and structured in a way that clearly explains to users that they’re choosing to bypass F-Droid’s checks if they activate it.`

In the settings I could explain well what the policy requires when user enabled new check update...

what do you think?

**Comment by ikanakova** [comment]
I don't think this permission is needed - it could be solved by the user selecting the folder to download the new version, and after downloading, the user could update manually in the Files app, for example.

**Comment by fast4x** [comment]
> > > check update feature include download and install package from GitHub
> > 
> > 
> > Thanks @fast4x! So the app now has a self-updater? You're aware that violates the inclusion policy of F-Droid and my repo as well (as in both cases, the additional screening applied by the repos is skipped)?
> > > when user click Update button
> > 
> > 
> > So it's kinda opt-in then (update check only happens when that button is pressed)? Is it clear to someone not having followed discussions here that updates then happen directly from Github – and are the implications (circumventing the extra security provided by F-Droid/my repo) clearly pointed out?
> > > All connections to GitHub or YT/YTM from app are in https, not have connection unencrypted.
> > 
> > 
> > Then why does the app have the `usesCleartextTraffic` set?
> > ```
> > <application android:theme="@7F0F011C" android:label="RiMusic" android:icon="@7F0D0001"
> >  android:name="it.vfsfitvnm.vimusic.MainApplication" android:configChanges="0x40007FFF"
> >  android:allowBackup="true" android:hardwareAccelerated="true" android:supportsRtl="true"
> >  android:banner="@7F0D0000" android:extractNativeLibs="true" android:usesCleartextTraffic="true"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe some dependency merged that in, and you could "override" it?
> 
> @IzzySoft Related to cleartext i had it enabled for debugging, I forgot to disable it. I'm afraid, disable it and rebuild.
> 
> Check update, check if there is an update on startup, if there is shows the user a message with an update button. Only if the user clicks Update start download and installation which still requires the user to confirm further.
> 
> This violates FDroid policy? Aigoo...

I checked my manifest, I had already removed usesCleartextTraffic, I found where it is reactivated. It’s the AppUpdate package that requires it in its manifest https://github.com/azhon/AppUpdate/blob/main/appupdate/src/main/AndroidManifest.xml Then I remove the package and go back to the previous check update method.

Thanks @IzzySoft for your control and support

**Comment by fast4x** [comment]
@DarkCrypt i'm afraid, i will develop new version without other permission request

**Comment by fast4x** [comment]
> > @fast4x looks like that permission list needs an update. Today my scanner reported additional things:
> > android.permission.REQUEST_INSTALL_PACKAGES: what for does a music player need to install apps?
> > usesCleartextTraffic: where are unencrypted connections established to?
> 
> I think this is related to issue #555

I removed package, i will develop new function without permission in the future

**Comment by ikanakova** [comment]
>>> @fast4x looks like that permission list needs an update. Today my scanner reported additional things:
>>> android.permission.REQUEST_INSTALL_PACKAGES: what for does a music player need to install apps?
>>> usesCleartextTraffic: where are unencrypted connections established to?
>> 
>> I think this is related to issue #555
> 
> I removed package, i will develop new function without permission in the future

I think it could be done by downloading a new version, but the user would have to manually install it themselves using the Files app, for example. Then this permission would not be needed.

**Comment by fast4x** [comment]
Using the system download service, the user should be able to manually launch the installation from the notification area. What do you think?

**Comment by fast4x** [comment]
![Screenshot_20240204_001144](https://github.com/fast4x/RiMusic/assets/1484476/15e4f62c-2ed3-48d0-82ee-a6fae50f60e8)
New check update with actions: open releases page and download with default policy without permission to install

**Comment by fast4x** [comment]
![Screenshot_20240204_002453](https://github.com/fast4x/RiMusic/assets/1484476/2c0bd541-4150-4429-a241-56819e267ac3)

With info for F-Droid users

**Comment by IzzySoft** [comment]
Yes, that looks fine! *If you wish* you could expand the last statement:

> F-Droid users can wait for the update from F-Droid, which usually arrives within a few hours with the IzzyOnDroid repo, and a few days with F-Droid.org's repo.

Icing on the cake, agreed :rofl:

**Comment by DarkCrypt** [comment]
@IzzySoft 
I understand what's all being said here, thank you for clarifying. I assume all apps that have this feature within your repo and F-Droid's repo have a similar "warning"? I'm probably wrong but I don't recall ever seeing this.

@fast4x 
Again, I do apologize for the headache. I'm glad to know this feature can still exist because of these changes. Your implementation looks great! I mainly made this request in hopes it'll work for Android TV. Thank you for everything you do!

**Comment by fast4x** [comment]
> Yes, that looks fine! *If you wish* you could expand the last statement:
> 
> > F-Droid users can wait for the update from F-Droid, which usually arrives within a few hours with the IzzyOnDroid repo, and a few days with F-Droid.org's repo.
> 
> Icing on the cake, agreed :rofl:

Yes! Will be add icing on the cake 😂

**Comment by fast4x** [comment]
![Screenshot_20240204_120701](https://github.com/fast4x/RiMusic/assets/1484476/c266f5b0-96f5-4e45-a37f-12e50407eb90)

I like it, very transparent and respectful to all types of users

**Comment by IzzySoft** [comment]
@DarkCrypt 

> I assume all apps that have this feature within your repo and F-Droid's repo have a similar "warning"? I'm probably wrong but I don't recall ever seeing this.

Yes. And yes; those additional checks were just implemented in January, which is why the results pop up only now. I reach out whenever an update triggers them. For existing apps, wherever you see those in the package details on the repo browser, be welcome to reach out to the corresponding project yourself to ask for clarification and ping me from the issue when clarifications have been made. I *might* make a "full scan" once reports from my updater have "calmed down" – but I cannot yet say when this might be.

**Comment by IzzySoft** [comment]
@fast4x Wonderful! :star_struck: So when that "popup" shows and I want "option 3" (wait for the F-Droid client to deal with the update), I tap that option (or the "back" key) to close it?

**Comment by fast4x** [comment]
> @fast4x Wonderful! :star_struck: So when that "popup" shows and I want "option 3" (wait for the F-Droid client to deal with the update), I tap that option (or the "back" key) to close it?

The popup by touching any point outside it. 🥳

**Comment by IzzySoft** [comment]
Thanks a lot! Couldn't resist this one then: https://floss.social/@IzzyOnDroid/111873690047110470 Thanks! So with the next release, `REQUEST_INSTALL_PACKAGES` will be gon again, right? Then this issue can be closed, as everything has been addressed.

**Comment by fast4x** [comment]
> Thanks a lot! Couldn't resist this one then: https://floss.social/@IzzyOnDroid/111873690047110470 Thanks! So with the next release, `REQUEST_INSTALL_PACKAGES` will be gon again, right? Then this issue can be closed, as everything has been addressed.
> 
> PS: The check itself is opt-in, too? Or does the app still check without consent? Because then it would need `NonFreeNet`, as it reaches out to a non-free network service.

Beautiful! Thank you very much for using RiMusic as an example for everyone. It’s a nice prize for me.

`REQUEST_INSTALL_PACKAGES` no longer exists, so in the next version there will be.

**Comment by fast4x** [comment]
Checking for the presence of the new version on GitHub is not opt-in, it was not before. Do I have to make it opt-in under the new rules?

P.s. The popup is shown only if there is a new version on GitHub... obviously after checking if a new version has been published.

**Comment by fast4x** [comment]
Maybe it is more correct to turn on/off the control from the settings, tonight I add it Will be disabled by default. User can enable or from settings.

Update also wiki page with this news.

**Comment by IzzySoft** [comment]
> Maybe it is more correct to turn on/off the control from the settings, tonight I add it Will be disabled by default. User can enable it from settings.

That sounds good! See [my comment from yesterday](https://github.com/fast4x/RiMusic/issues/400#issuecomment-1925303222), quote:

> Suggestion: have a new setting item for that, initialized with `NULL`. When starting the app and that field is set to `NULL`, inform about the updater mentioning source and implication and ask whether to have in enabled or disabled. Then remember that setting (and act accordingly). The option in settings of course remains accessible should one have a change of mind.

Thanks to the latest updates, that can be shortened to something like "Check at Github for updates?" – though it cannot hurt to add "When an update is available, you will be asked if you want to install it. There won't be automated installs „behind your back“."

**Comment by fast4x** [comment]
![Screenshot_20240205_114923](https://github.com/fast4x/RiMusic/assets/1484476/8fc9444f-4e5a-4c52-9d98-64755e37f915)
Implemented just available from version 0.6.21

**Comment by IzzySoft** [comment]
Well, one could lament that the "Don't enable" isn't visible to be a button ("dark patterns", uuh!) – so maybe give it some background (e.g. grayish; I'd not choose reddish for the very same reason: it looks like "danger" then).

Just a suggestion, not a demand of course :wink:

Thanks a lot for following this through!! :heart_eyes:

**Comment by fast4x** [comment]
![Screenshot_20240205_151410](https://github.com/fast4x/RiMusic/assets/1484476/83bd23ce-7bc3-4d50-8814-b737d1f3220c)
Nice!

**Comment by fast4x** [comment]
![Screenshot_20240205_152302](https://github.com/fast4x/RiMusic/assets/1484476/03c2ad12-9192-4304-883c-187f901941e1)
This is very nice

**Comment by IzzySoft** [comment]
I like both. And if someone asks: Green is "on", gray is "off" – so no "dark pattern" there :smile:

**Comment by fast4x** [comment]
ahahaha ok LOL

**Comment by IzzySoft** [comment]
Though having both buttons green would be more fun. A little "confusion" would ensure people really read those few lines and do not blindly tap the green button. Standing between two of them, they might get an idea how [Buridan's ass](https://en.wikipedia.org/wiki/Buridan%27s_ass) felt :rofl:

**Comment by fast4x** [comment]
Ok both green so they read well! The experience of [buridan ](https://en.wikipedia.org/wiki/Buridan%27s_ass) teaches 🤣

> ![Screenshot_20240205_151410](https://private-user-images.githubusercontent.com/1484476/302318564-83bd23ce-7bc3-4d50-8814-b737d1f3220c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNDg2ODksIm5iZiI6MTcwNzE0ODM4OSwicGF0aCI6Ii8xNDg0NDc2LzMwMjMxODU2NC04M2JkMjNjZS03YmMzLTRkNTAtODgxNC1iNzM3ZDFmMzIyMGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDVUMTU1MzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDkyY2Y3YmQ3NjkyZTU5OGVhNGI5Y2QyYWY5MGYzNTAxNmE3YTZmMDQ1MWIwYWViM2UxNDUxYmFjNWRiYTQzNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.JYuzmoHbzsolb9oFDVEuyGtLGXpQFYT8EtIbu9MAwaQ) Nice!

Ok both green so they read well! The experience of [buridan ](https://en.wikipedia.org/wiki/Buridan%27s_ass) teaches 🤣

**Comment by IzzySoft** [comment]
! repo/it.fast4x.rimusic_17.apk declares sensitive permission(s): android.permission.READ_EXTERNAL_STORAGE

What shall we put for that then? I guess it's needed to play music stored "somewhere else" (like an SD card, or an attached thumb drive)?

**Comment by fast4x** [comment]
It's needed up to Android 12 for play local Song, without not work.

**Comment by fast4x** [comment]
>     ! repo/it.fast4x.rimusic_17.apk declares sensitive permission(s): android.permission.READ_EXTERNAL_STORAGE
> 
> What shall we put for that then? I guess it's needed to play music stored "somewhere else" (like an SD card, or an attached thumb drive)?

https://github.com/fast4x/RiMusic/issues/646

**Comment by IzzySoft** [comment]
Thanks! Added.

PS: Are you OK if I use some of the above screenshots as examples to illustrate how to deal with "self-updates"?

**Comment by fast4x** [comment]
No problem I’m honored

**Comment by IzzySoft** [comment]
Then expect them to show up in unexpected places :rofl: Thanks!

**Comment by IzzySoft** [comment]
First occurrence just happened. While posting the pics there I found a minor "idea for improvement": the details don't mention the implications. Would be real nice if you could expand the description in the first place a little:

> Check at Github for updates?
>
> When an update is available, you will be asked if you want to install it. There won't be automated installs, behind your back. **But these updates would not go through the additional screening performed at e.g. IzzyOnDroid or F-Droid.**

No need to make the addition bold – I just did that here so you see what was changed/added. I think it's neutral enough to not imply you'd deliver something weird – it's just "additional screening" as "more eyes" see/catch more things. Can never hurt to have additional screenings – especially when they come for free, right? :wink:

And apologies that I put IoD in front of F-Droid here. The reason is IoD has several checks in place F-Droid lacks, and does not seem willing to implement.

**Comment by fast4x** [comment]
I like the extra detail so I add it with pleasure.

**Comment by IzzySoft** [comment]
:star_struck: Please let me have a screenshot then to replace the one I'm currently using :pray:

**Comment by fast4x** [comment]
Yes, tomorrow

**Comment by fast4x** [comment]
@IzzySoft screenshot ready
![Screenshot_20240211_162007](https://github.com/fast4x/RiMusic/assets/1484476/27af569c-79a1-4882-b245-bc719e9aa7e7)

**Comment by IzzySoft** [comment]
And the 2 green buttons, fantastic! I'm still ROFLMAO imagining the first confused looks on faces: "WTF? And which one shall I push?" – with the result they're really **reading** the text. Great work, @fast4x – I'll make sure to spread the word!

**Comment by fast4x** [comment]
Great! 😘

**Comment by fast4x** [comment]
@IzzySoft Unfortunately I have to reintroduce the permission android.permission.WRITE_EXTERNAL_STORAGE because the download does not work until Android 10.

**Comment by IzzySoft** [comment]
Thanks for the advance notice! I'll then know what to do when the warning comes in (will add it to the allow-list with the proper explanation straight away then).

**Comment by fast4x** [comment]
Thanks you.

**Comment by IzzySoft** [comment]
Ah, the notification that I wanted to leave a final note here. You got another prominent place for the screenshots:

[Ramping up security: additional APK checks are in place with the IzzyOnDroid repo](https://android.izzysoft.de/articles/named/iod-scan-apkchecks)

:smile:

**Comment by fast4x** [comment]
Wow it’s very nice to read Rimusic as an example in your note. ,😊😅😜🥳💕

**Comment by IzzySoft** [comment]
You gave me a very nice example to use, and the permission to use it – so I did. Great work at your side, so honor whom honor is due :smile:

**Comment by IzzySoft** [comment]
After this nice success story, want to add some icing on the cake and put a cherry on top? There was another check added a while ago, which for RiMusic says:

    ! repo/it.fast4x.rimusic_30.apk contains signature block blobs: 0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)

It's an easy task:

```gradle
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains.

Thanks in advance, @fast4x!

**Comment by fast4x** [comment]
Thanks I'll add code in gradle soon. Thanks you @IzzySoft

**Comment by IzzySoft** [comment]
Wonderful, thank you again!

**Comment by fast4x** [comment]
Added, available from version 0.6.33
Thanks @IzzySoft

**Comment by IzzySoft** [comment]
Thanks! :star_struck:
